### PR TITLE
[DoctrineBridge] Fix LockStoreSchemaListener not working properly with iterable objects

### LINKED ISSUE #54406: [DoctrineBridge] LockStoreSchemaListener not working properly
### Symfony version(s) affected

6.4.4

### Description

After installing `symfony/lock` and configuring it with `DoctrineDbalStore`, I noticed that invoking `doctrine:migrations:diff` command would generate a migration issuing a `DROP` of the autogenerated `lock_keys` table.

After investigating, I found out the issue is [line 33 of the listener](https://github.com/symfony/symfony/blob/7.1/src/Symfony/Bridge/Doctrine/SchemaListener/LockStoreSchemaListener.php#L33). The `$stores` property receives a `RewindableGenerator`, which is then passed directly to an `ArrayIterator`. This produces an invalid iterator, skipping the whole following `while` loop.

### How to reproduce

1. Configure `symfony/lock` with a `DoctrineDbalStore`.
2. Generate the `lock_keys` table, either by using the `LockFactory` or by any other means mentioned in the [docs](https://symfony.com/doc/current/components/lock.html#doctrinedbalstore).
3. Run the `doctrine:migrations:diff` command.
4. See that the the schema for the table is not properly configured, resulting in a migration instead of the expected `No changes detected in your mapping information.` message.

### Possible Solution

Since the `$stores` property is already typed as `iterable`, it would seem replacing the loop with a `foreach` iterating directly on the property would be the best solution.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by barton-webwings**
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | yes
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix #54406
| License       | MIT

### CODE REVIEW

**Comment by MatTheCat** [comment]
I think you can simply revert #50761:

- `ArrayIterator` won’t work as expected when passed a traversable (this is why `testPostGenerateSchemaWithInvalidLockStore` passes: its iterator won’t yield anything. You can be sure because its generator uses `$this` while being `static`).
- Exceptions thrown inside a generator will appear when calling `valid` or `next`, not `current`.
- An exception thrown inside a generator would close it, even if caught when iterating; this could result in other stores being silently ignored.
- `LogicException`s are not supposed to be caught.

**Comment by alexandre-daubois** [comment]
Indeed my solution doesn't work as expected. I don't think that reverting the PR is the best solution, because it would fix a bug but reintroduce one in the codebase.

I'm on my phone right now but I can have a look at it tomorrow to find/have a look for a proper solution, if you need help 🙂

**Comment by barton-webwings** [comment]
I think there could be a better solution to the original issue https://github.com/symfony/symfony/pull/50761.

Personally, I would take a similar path to Monolog's `NullHandler` or Mailer's `NullTransport` and introduce a `NullStore` and a corresponding case to the `StoreFactory`. The DSN could look something like `LOCK_STORE=null://`. 

This way we would avoid the `InvalidArgumentException`, which I agree is a legitimate exception, but preserve the option to effectively disable the Lock component.

**Comment by alexandre-daubois** [comment]
Seems to be a great solution indeed, I like it!

**Comment by barton-webwings** [comment]
I pushed a draft of my proposed solution into this PR, hopefully I didn't miss anything.

Since the `StoreFactory` already has cases which are not strictly a dsn, I went with a simple `null`, which would theoretically work with existing cases of `LOCK_STORE=null`.

**Review by alexandre-daubois — APPROVED** [review]
Seems like a clean solution like this 😊

**Comment by nicolas-grekas** [comment]
So, that's a new feature now, isn't it?

**Comment by Toflar** [comment]
Not really a new feature. The current implementation is broken. It will not add the required schema as described in https://github.com/symfony/symfony/issues/54406. Just faced the same issue and this PR fixes it :)

**Comment by MatTheCat** [comment]
The `NullStore` is a new feature.

#50761 revert is a fix.

File: src/Symfony/Component/Lock/Store/NullStore.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+use Symfony\Component\Lock\BlockingSharedLockStoreInterface;
+use Symfony\Component\Lock\Key;
+
+/**
+ * NullStore is a PersistingStoreInterface implementation which discards all operations.
+ *
+ * @author Pavel Bartoň <barton@webwings.cz>
+ */
+class NullStore implements BlockingSharedLockStoreInterface
```
Comment (xabbuh) [inline src/Symfony/Component/Lock/Store/NullStore.php:22]: Introducing this class (as well as the changes to the `StoreFactory`) must be moved into a separate PR targeting the `7.2` branch. That's a new feature.

File: src/Symfony/Bridge/Doctrine/Tests/SchemaListener/LockStoreSchemaListenerTest.php [hunk 2]
```diff
@@ -40,20 +39,4 @@ public function testPostGenerateSchemaLockPdo()
         $subscriber = new LockStoreSchemaListener([$lockStore]);
         $subscriber->postGenerateSchema($event);
     }
-
-    public function testPostGenerateSchemaWithInvalidLockStore()
```
Comment (xabbuh) [inline src/Symfony/Bridge/Doctrine/Tests/SchemaListener/LockStoreSchemaListenerTest.php:44]: this test should be kept as fixing the issue #50761 tried to fix is legitimate

File: src/Symfony/Bridge/Doctrine/SchemaListener/LockStoreSchemaListener.php [hunk 3]
```diff
@@ -30,20 +29,12 @@ public function postGenerateSchema(GenerateSchemaEventArgs $event): void
...
-        $storesIterator = new \ArrayIterator($this->stores);
-        while ($storesIterator->valid()) {
-            try {
-                $store = $storesIterator->current();
-                if (!$store instanceof DoctrineDbalStore) {
-                    continue;
-                }
-
-                $store->configureSchema($event->getSchema(), $this->getIsSameDatabaseChecker($connection));
-            } catch (InvalidArgumentException) {
```
Comment (xabbuh) [inline src/Symfony/Bridge/Doctrine/SchemaListener/LockStoreSchemaListener.php:42]: we still need catch the exception to not reintroduce the bug that #50761 tried to fix

**Review by xabbuh — CHANGES_REQUESTED** [review]

**Comment by MatTheCat** [comment]
@xabbuh there was in fact no bug prior to #50761: it is normal an exception is thrown when giving the `StoreFactory` an invalid DSN, and it is outlined by the fact said exception extends `LogicException`, which is not supposed to be caught.

**Comment by uwej711** [comment]
OK, so where do we go from here? Keep the current bug in 6.4, revert the commit that causes the bug and having the issue it tried to fix again or something else?

**Comment by xabbuh** [comment]
I would be okay with both solutions. Either reverting #50671 or switching to `foreach`). But the new `null` schema cannot be merged into 6.4 and must be moved into a PR targeting 7.2.

**Comment by barton-webwings** [comment]
I am closing this PR, as it is superseded by https://github.com/symfony/symfony/pull/57944.
I am actually only interested in the bugfix part of this issue, so we can remove the workaround after update.
If someone wants to make a PR for the new feature introduced, feel free.

**Pull request closed without merging**